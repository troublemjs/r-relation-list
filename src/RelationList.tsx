import React, { useLayoutEffect, useRef, useState } from 'react';
import classNames from 'classnames';

import Item from './Item';

const EMPTY_DATA: never[] = [];

export interface RelationListProps<T> {
  renderItem?: (record: T, index: number) => React.ReactNode;
  title?: React.ReactNode;
  dataSource?: T[];
  rowKey?: ((item: T) => string) | string;
  empty?: React.ReactNode;
  prefixCls?: string;
  className?: string;
  children?: React.ReactNode;
  style?: React.CSSProperties;
  bordered?: boolean;
  /** 是否展示分割线 */
  split?: boolean;
  header?: React.ReactNode;
  footer?: React.ReactNode;
}
const RelationList = <RecordType extends unknown>(
  props: RelationListProps<RecordType>,
) => {
  const {
    prefixCls,
    title,
    renderItem,
    dataSource,
    empty,
    className,
    rowKey,
    bordered,
    split,
    header,
    footer,
    children,
    ...restProps
  } = props;
  const mergedData = dataSource || EMPTY_DATA;
  const ulRef = useRef<HTMLUListElement>(null);
  const [contentLineWidth, setContentLineWidth] = useState<number>(0);

  useLayoutEffect(() => {
    const ulNode = ulRef.current;
    const ulHeight = ulNode?.clientHeight ?? 0;
    const heithg0th = ulNode?.firstElementChild?.clientHeight ?? 0;

    const contentLineHeight = ulHeight - heithg0th;
    console.log(contentLineHeight);

    setContentLineWidth(contentLineHeight);
  }, [mergedData]);

  const keys: { [key: string]: string } = {};

  const renderInnerItem = (item: RecordType, idx: number) => {
    if (!renderItem) return null;

    let key;
    if (typeof rowKey === 'function') {
      key = rowKey!(item);
    } else if (typeof rowKey === 'string') {
      key = rowKey;
    } else {
      key = (item as any).key;
    }

    if (!key) {
      key = `list-item-${idx}`;
    }
    keys[idx] = key;

    return renderItem(item, idx);
  };
  // if (mergedData.length === 0) return empty;

  let childrenContent: React.ReactNode;
  if (mergedData.length > 0) {
    const items = mergedData.map(renderInnerItem);
    const newItems = React.Children.map(items, (child, idx) => (
      <li key={keys[idx]} className={`${prefixCls}-item-wrap`}>
        <div className={`${prefixCls}-item-prefix-line`} />
        {child}
      </li>
    ));
    childrenContent = (
      <ul ref={ulRef} className={`${prefixCls}-items`}>
        {newItems}
      </ul>
    );
  }

  const shouldContent = Boolean(header || footer);
  const Content = shouldContent ? 'div' : React.Fragment;

  const classString = classNames(
    prefixCls,
    {
      [`${prefixCls}-split`]: split,
      [`${prefixCls}-bordered`]: bordered,
      [`${prefixCls}-content`]: !shouldContent,
    },
    className,
  );

  return (
    <div className={classString} {...restProps}>
      {header && <div className={`${prefixCls}-header`}>{header}</div>}
      <Content className={`${prefixCls}-content`}>
        <span className={`${prefixCls}-title`}>{title}</span>
        {childrenContent}
        {children}
      </Content>
      {footer && <div className={`${prefixCls}-footer`}>{footer}</div>}
      {/* <div
        className={`${prefixCls}-parent-line`}
        style={{ height: firstHeight }}
      /> */}
      {/* <ul ref={ulRef} className={`${prefixCls}-content`}> */}
      {/* {mergedData?.map((item, idx) => (
          <li key={idx} className={`${prefixCls}-item`}>
            <div className={`${prefixCls}-item-vertical-line`} />
            {renderItem?.(item, idx) ?? `${item}`}
          </li>
        ))} */}
      {/* </ul> */}
    </div>
  );
};

RelationList.Item = Item;

RelationList.defaultProps = {
  prefixCls: 'relation-list',
  dataSource: [],
  empty: '无数据',
  bordered: true,
  split: true,
};

export default RelationList;
