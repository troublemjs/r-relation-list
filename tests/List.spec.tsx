import React from 'react';
import { mount } from 'enzyme';

import RelationList from '../src';

describe('RelationList.Basic', () => {
  const data = [1, 2, 3, 4, 5];
  const createList = (props) => (
    <RelationList<typeof data[number]>
      dataSource={data}
      renderItem={(item) => item}
      {...props}
    />
  );

  describe('renders correctly', () => {
    it('basic', () => {
      const wrapper = mount(
        createList({
          prefixCls: 'test-prefix',
          className: 'test-class-name',
        }),
      );
      expect(wrapper.render()).toMatchSnapshot();
    });

    it('header', () => {
      const wrapper = mount(
        createList({
          header: 'test-header',
        }),
      );
      expect(wrapper.render()).toMatchSnapshot();
    });

    it('footer', () => {
      const wrapper = mount(
        createList({
          footer: 'test-footer',
        }),
      );
      expect(wrapper.render()).toMatchSnapshot();
    });
  });

  describe('renders empty correctly', () => {
    describe('renders ReactNode correctly', () => {
      it('has title', () => {
        const wrapper = mount(
          createList({ title: 'Title', dataSource: [], empty: 'No data' }),
        );

        expect(wrapper.find('.relation-list-title').text()).toEqual('Title');

        expect(wrapper.find('.relation-list').children().last().text()).toEqual(
          'No data',
        );
      });

      it('no title', () => {
        const wrapper = mount(createList({ dataSource: [], empty: 'No data' }));
        expect(wrapper.text()).toEqual('No data');
      });
    });

    it('effect update', () => {
      const App = () => {
        const [empty, setEmpty] = React.useState('No data');
        React.useEffect(() => {
          setEmpty('暂无数据');
        }, []);
        return <RelationList empty={empty} />;
      };
      const wrapper = mount(<App />);
      wrapper.update();
      expect(wrapper.find('.relation-list').text()).toEqual('暂无数据');
    });
  });
});
