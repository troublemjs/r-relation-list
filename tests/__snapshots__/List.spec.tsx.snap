// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RelationList.Basic renders correctly basic 1`] = `
<div
  class="test-prefix test-prefix-split test-prefix-bordered test-prefix-content test-class-name"
>
  <ul
    class="test-prefix-items"
  >
    <li
      class="test-prefix-item-wrap"
    >
      <div
        class="test-prefix-item-prefix-line"
      />
      1
    </li>
    <li
      class="test-prefix-item-wrap"
    >
      <div
        class="test-prefix-item-prefix-line"
      />
      2
    </li>
    <li
      class="test-prefix-item-wrap"
    >
      <div
        class="test-prefix-item-prefix-line"
      />
      3
    </li>
    <li
      class="test-prefix-item-wrap"
    >
      <div
        class="test-prefix-item-prefix-line"
      />
      4
    </li>
    <li
      class="test-prefix-item-wrap"
    >
      <div
        class="test-prefix-item-prefix-line"
      />
      5
    </li>
  </ul>
</div>
`;

exports[`RelationList.Basic renders correctly footer 1`] = `
<div
  class="relation-list relation-list-split relation-list-bordered"
>
  <div
    class="relation-list-content"
  >
    <ul
      class="relation-list-items"
    >
      <li
        class="relation-list-item-wrap"
      >
        <div
          class="relation-list-item-prefix-line"
        />
        1
      </li>
      <li
        class="relation-list-item-wrap"
      >
        <div
          class="relation-list-item-prefix-line"
        />
        2
      </li>
      <li
        class="relation-list-item-wrap"
      >
        <div
          class="relation-list-item-prefix-line"
        />
        3
      </li>
      <li
        class="relation-list-item-wrap"
      >
        <div
          class="relation-list-item-prefix-line"
        />
        4
      </li>
      <li
        class="relation-list-item-wrap"
      >
        <div
          class="relation-list-item-prefix-line"
        />
        5
      </li>
    </ul>
  </div>
  <div
    class="relation-list-footer"
  >
    test-footer
  </div>
</div>
`;

exports[`RelationList.Basic renders correctly header 1`] = `
<div
  class="relation-list relation-list-split relation-list-bordered"
>
  <div
    class="relation-list-header"
  >
    test-header
  </div>
  <div
    class="relation-list-content"
  >
    <ul
      class="relation-list-items"
    >
      <li
        class="relation-list-item-wrap"
      >
        <div
          class="relation-list-item-prefix-line"
        />
        1
      </li>
      <li
        class="relation-list-item-wrap"
      >
        <div
          class="relation-list-item-prefix-line"
        />
        2
      </li>
      <li
        class="relation-list-item-wrap"
      >
        <div
          class="relation-list-item-prefix-line"
        />
        3
      </li>
      <li
        class="relation-list-item-wrap"
      >
        <div
          class="relation-list-item-prefix-line"
        />
        4
      </li>
      <li
        class="relation-list-item-wrap"
      >
        <div
          class="relation-list-item-prefix-line"
        />
        5
      </li>
    </ul>
  </div>
</div>
`;
